rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+        other_labels = []  # slots
+        for feature_name in input_feature_map.keys():
+            if (feature_name.startswith('intent_')
+                    # include entity as user input
+                    or feature_name.startswith('entity_')):
+                user_labels.append(feature_name)
+            elif feature_name.startswith('prev_'):
+                bot_labels.append(feature_name[len('prev_'):])
+            else:
+                other_labels.append(feature_name)
+
+        return user_labels, bot_labels, other_labels
+
+    @staticmethod
+    def _create_label_dict(labels):
+        """Create label dictionary"""
+        label_dict = {}
+        for label in labels:
+            if label not in label_dict:
+                label_dict[label] = len(label_dict)
+        return label_dict
+
+    @staticmethod
+    def _create_label_token_dict(labels, split_symbol='_'):
+        """Create label dictionary"""
+        label_token_dict = {}
+        for label in labels:
+            for t in label.split(split_symbol):
+                if t not in label_token_dict:
+                    label_token_dict[t] = len(label_token_dict)
+        return label_token_dict
+
+    def create_helpers(self, domain):
+
+        self.labels = self._parse_feature_map(domain.input_feature_map)
+        user_labels, bot_labels, other_labels = self.labels
+
+        self.bot_actions = self._create_label_dict(bot_labels)
+
+        if not self.share_vocab:
+            self.bot_vocab = self._create_label_token_dict(bot_labels,
+                                                           self.split_symbol)
+            self.user_vocab = self._create_label_token_dict(user_labels,
+                                                            self.split_symbol)
+        else:
+            self.bot_vocab = self._create_label_token_dict(bot_labels + user_labels)
+            self.user_vocab = self.bot_vocab
+
+        self.additional_features = other_labels
+
+    def encode(self, active_features, domain):
+
+        user_labels, bot_labels, other_labels = self.labels
+
+        num_features = (len(self.user_vocab) +
+                        len(self.bot_vocab) +
+                        len(self.additional_features))
+
+        if active_features is None or None in active_features:
+            return np.ones(num_features, dtype=int) * -1
+
+        used_features = np.zeros(num_features, dtype=int)
+        for feature_name, prob in active_features.items():
+
+            if feature_name in user_labels:
+                if 'prev_action_listen' not in active_features:
+                    # we predict next action from bot action
+                    # feature_name = 'intent_listen'
+                    used_features[:len(self.user_vocab)] = 0
+                else:
+                    for t in feature_name.split(self.split_symbol):
+                        used_features[self.user_vocab[t]] += 1
+
+            elif feature_name[len('prev_'):] in bot_labels:
```

Comment:
I think we should change the structure of the collected features and making them objectes (containing name, type and value) - but let's not do it as part of this pr.

Reply:
yes, good idea, let's do it together with refactoring domain